### DISCUSSION TITLE
Is there a way to make a uv_stream_t from an already accepted tcp connection?

### DISCUSSION THREAD

**Discussion by upsampled** [libuv/libuv#3385] [Q&A] [answered]
I am being passed an already accepted tcp connection file descriptor from another program, is there any way to turn it into a uv_stream_t?

**Comment by bnoordhuis** [comment]
uv_tcp_open()?

**Reply by upsampled** [reply]
I tried that but it did not appear to work. I think I need to go directly to a uv_stream_t

**Comment by bnoordhuis** [comment]
A uv_tcp_t is a kind of uv_stream_t. Can you elaborate on what you tried and what didn't work?

**Comment by upsampled** [comment]
Ok, essentially the application is being socket activated where stdin/stdout is actually an accepted connection. I want to use this tcp connection exactly like i used a `uv_tty_t` for normal execution. I do the following do initialize.

```c
error_t *uv_term_tcp(uv_loop_t *loop, m_config_t *cfg, m_ctx_t *ctx)
{
    int rv;


    rv = uv_tcp_init(loop, &ctx->term.tcpuv);
    if (rv < 0)
    {
        sprintf(ctx->term.err.errmsg, "error initializing term tcp ");
        return &ctx->term.err;
    }

    if (strncmp("STDIN", cfg->ttydev, TTYDEVNAME) == 0)
    {
        log_info("term is set to stdin");
        rv = uv_tcp_open(&ctx->term.tcpuv, STDIN_FILENO);
        if (rv < 0)
        {
            sprintf(ctx->term.err.errmsg, "error opening term tcp ");
            return &ctx->term.err;
        }
        ctx->term.tcpuv.data = ctx;
        return NULL;
    }
    sprintf(ctx->term.err.errmsg, "error initializing term tcp, unknonw input ");
    return &ctx->term.err;
}
```

I then treat the uv_tcp_t just like i do the t_tty_t and place it in `uv_read_start`. I get the following error at launch:

```
src/unix/stream.c:1402: uv_write2: Assertion `(stream->type == UV_TCP || stream->type == UV_NAMED_PIPE || stream->type == UV_TTY) && "uv_write (unix) does not yet support other types of streams"' failed.`
```

**Comment by bnoordhuis** [comment, marked as answer]
Your code should work, I don't see anything obviously wrong with it.

My best guess is that something clobbers `ctx->term.tcpuv` after the function returns. What does its contents look like when you hit the assert in gdb?

**Reply by upsampled** [reply]
very possible, just knowing that `uv_tcp_open` should handle this greatly scopes the problem
